### PR TITLE
Auto-detect Github Actions CI and activate github logger accordingly

### PR DESCRIPTION
**Pull request by mfn**
This PR:

- [x] Adds new feature
- [x] Covered by tests
- [x] Doc PR: https://github.com/infection/site/pull/231

Fixes https://github.com/infection/infection/issues/1560

---
With this PR, the `--logger-github` flag is not necessary when running in Github Actions anymore. It's still supported:
- doesn't break existing scripts
- can be used to force-enable this particular logger in non-Github Action environments if desired

When running on Github Actions, however, it's not possible to "disable" this logger; as long as the Github Actions CI is correctly detected according to OndraM/ci-detector, it will always be active. This is deliberately done to improve the out of the box experience with infection.

### CODE REVIEW

**Review by maks-rafalko — APPROVED** [review]
:+1:

**Comment by maks-rafalko** [comment]
Build failure

> The minimum required MSI percentage should be 71.87%, but actual is 71.86%. Improve your tests!  

it's not your fault, but comes from `master`. You can however change this line https://github.com/infection/infection/blob/master/.github/workflows/mt.yaml#L14 and update it to `71.86` so builds will pass.

Anyway, I'm happy to merge, please update the docs.

**Comment by mfn** [comment]
Thanks! I've pushed three more changes:
- mention the auto-detect also on the cli help itself
- lower the msi as suggested
- "eat our own dogfood" and remove the flag from the workflow itself
- further: docs PR at https://github.com/infection/site/pull/231

**Comment by maks-rafalko** [comment]
looks like our end-to-end tests are broken now (https://github.com/infection/infection/runs/4787795564?check_suite_focus=true#step:12:5)

I suspect this is because tests, executed from PHPUnit, started producing GitHub warning messages (https://github.com/infection/infection/runs/4787795564?check_suite_focus=true#step:12:27) and this wasn't exptected from PHPUnit.

I have a feeling that we need to disable github logger here. Since it's being enabled automatically, probably we need to make it possible to pass `--logger-github=false` here:

https://github.com/infection/infection/blob/07490ad852ddc868055ae47f19f2c5bc5c0e4a9e/tests/phpunit/Console/E2ETest.php#L351

**Comment by mfn** [comment]
@maks-rafalko what a gotcha 😅

I've pushed a commit to support `--logger-github=false`, [from the commit](https://github.com/infection/infection/pull/1645/commits/788a43203c3dee31b06af179abdac3aa42a0b30e)

> TL;DR:
> - not provided: auto-detected
> - `--logger-github` : enabled
> - `--logger-github=true` : enabled
> - `--logger-github=false` : disabled

I hope this is what you had in mind; I had to change it to nullable etc.

Though I've to trouble to figure out where to explicitly put `--logger-github=false`; somewhere in `.github/workflows/ci.yaml` I presume?

File: src/Configuration/ConfigurationFactory.php [hunk 1]
```diff
@@ -328,4 +333,19 @@ private function retrieveLogs(Logs $logs, bool $useGitHubLogger, ?string $htmlLo
...
+
+            if ($ci->getCiName() === CiDetector::CI_GITHUB_ACTIONS) {
+                return true;
+            }
+        } catch (CiNotDetectedException $e) {
+            // deliberately empty
+        }
+
+        return false;
+    }
```
Comment (localheinz) [inline src/Configuration/ConfigurationFactory.php:350]:
```suggestion
    private function detectCiGithubActions(): bool
    {
        try {
            $ci = $this->ciDetector->detect();
        } catch (CiNotDetectedException $e) {
            return false;
        }
        
        if ($ci->getCiName() !== CiDetector::CI_GITHUB_ACTIONS) {
            return false;
        }

        return true;
    }
```

File: src/Command/RunCommand.php [hunk 2]
```diff
@@ -437,6 +438,24 @@ private function createContainer(IO $io, LoggerInterface $logger): Container
...
+        } elseif ($useGitHubLogger === null) {
+            $useGitHubLogger = true;
+        } elseif ($useGitHubLogger === 'true') {
+            $useGitHubLogger = true;
+        } elseif ($useGitHubLogger === 'false') {
+            $useGitHubLogger = false;
+        } else {
+            throw new InvalidArgumentException(
+                sprintf('Cannot pass "%s" to "--%s": only "true", "false" or no argument is supported', $useGitHubLogger, self::OPTION_LOGGER_GITHUB)
+            );
```
Comment (localheinz) [inline src/Command/RunCommand.php:456]:
```suggestion
            throw new InvalidArgumentException(sprintf(
                'Cannot pass "%s" to "--%s": only "true", "false" or no argument is supported', 
                $useGitHubLogger, 
                self::OPTION_LOGGER_GITHUB
            ));
```

**Comment by maks-rafalko** [comment]
@mfn could you please rebase?

**Comment by mfn** [comment]
> @mfn could you please rebase?

Done and force-pushed!

File: src/Configuration/ConfigurationFactory.php (hunk 1 above)
Comment (maks-rafalko) [reply src/Configuration/ConfigurationFactory.php]: :heavy_check_mark:

File: src/Command/RunCommand.php (hunk 2 above)
Comment (maks-rafalko) [reply src/Command/RunCommand.php]: :heavy_check_mark:

**Comment by maks-rafalko** [comment]
I've rebased and fixed e2e tests.

Thanks for this feature @mfn!

**Comment by mfn** [comment]
Thank you for your time too 🙏🏼

**Comment by maks-rafalko** [comment]
Released: https://github.com/infection/infection/releases/tag/0.26.11
